### PR TITLE
[EFIP-5] {Permit Whitelist, Transfer Blacklist}

### PR DESCRIPTION
**Pull request by jtfirek**
- Implements [EFIP-5](https://github.com/etherfi-protocol/smart-contracts/blob/master/proposals/EFIP-5.md) functionality and tests
- Makes comment style consistent between `eETH` and `weETH`

### CODE REVIEW

File: src/EETH.sol [hunk 1]
```diff
@@ -15,12 +15,18 @@ import "./interfaces/ILiquidityPool.sol";
...
+    //---------------------------------  STATE-VARIABLES  ----------------------------------
+    //--------------------------------------------------------------------------------------
+    
     ILiquidityPool public liquidityPool;
 
     uint256 public totalShares;
     mapping (address => uint256) public shares;
     mapping (address => mapping (address => uint256)) public allowances;
     mapping (address => CountersUpgradeable.Counter) private _nonces;
+    mapping (address => bool) public whitelistedSpender;
```
Comment (jtfirek) [inline src/EETH.sol:28]: @seongyun-ko, Could you confirm my reasoning here:

Becausing constant and immutable variables are stored directly in the bytecode, I can place the new mapping above these values without having to worry about a storage collision

Comment (seongyun-ko) [reply src/EETH.sol:28]: ho, sounds right :) you can always verify it by checking the storage layout of this contract before/after your update

File: src/EETH.sol [hunk 2]
```diff
@@ -139,8 +153,13 @@ contract EETH is IERC20Upgradeable, UUPSUpgradeable, OwnableUpgradeable, IERC20P
         _approve(owner, spender, value);
     }
 
-    // [INTERNAL FUNCTIONS] 
+    //--------------------------------------------------------------------------------------
+    //-------------------------------  INTERNAL FUNCTIONS  ---------------------------------
+    //--------------------------------------------------------------------------------------
+
     function _transfer(address _sender, address _recipient, uint256 _amount) internal {
+        require(blacklistedRecipient[_recipient] == false, "eETH: recipient blacklisted");
```
Comment (seongyun-ko) [inline src/EETH.sol:161]: According to the spec, the check happens  both on sender and recipient; 

```
require(!blacklisted[_sender] || !blacklisted[_recipient], "EETH: blacklisted address");
```

Comment (seongyun-ko) [reply src/EETH.sol:161]: Q2) how much gas cost does it add?

File: src/WeETH.sol [hunk 3]
```diff
@@ -76,14 +82,64 @@ contract WeETH is ERC20Upgradeable, UUPSUpgradeable, OwnableUpgradeable, ERC20Pe
...
     ) internal override onlyOwner {}
 
+
+    /// @notice Require the recipient to not be blacklisted before calling {ERC20Upgradeable-_transfer}
+    function _transfer(
+        address from,
+        address to,
+        uint256 amount
+    ) internal override {
+        require(blacklistedRecipient[to] == false, "weETH: recipient blacklisted");
```
Comment (seongyun-ko) [inline src/WeETH.sol:117]: the same here

File: test/TestSetup.sol [hunk 4]
```diff
@@ -703,6 +703,12 @@ contract TestSetup is Test {
         _initializeEtherFiAdmin();
 
         admin = alice;
+
+        // weETH Instance must be on eETH whitelist for `weETH.wrapWithPermit` to function as expected
```
Comment (seongyun-ko) [inline test/TestSetup.sol:707]: kk

**Review by seongyun-ko — CHANGES_REQUESTED** [review]
THX!

File: src/EETH.sol (hunk 2 above)
Comment (jtfirek) [reply src/EETH.sol:161]: Initial Gas Cost: 49948
Gas Cost with check only on `to`: 52161
Gas Cost if we add checks on both `to` and `from`: 54382

File: src/WeETH.sol (hunk 3 above)
Comment (jtfirek) [reply src/WeETH.sol:117]: Initial Gas Cost: 29848
Gas Cost with check only on `to`: 32100
Gas Cost if we add checks on both `to` and `from`: 34321

**Comment by jtfirek** [comment]
@solipsis @seongyun-ko up to you guys if you think the extra 2000 gas on the `SLOAD` is worth it to apply the sender as well

File: src/EETH.sol (hunk 2 above)
Comment (seongyun-ko) [reply src/EETH.sol]: thx that is good to know.

2100 is due to the cold SLOAD :)
https://hackmd.io/@fvictorio/gas-costs-after-berlin#SLOAD-after-Berlin

File: src/WeETH.sol (hunk 3 above)
Comment (seongyun-ko) [reply src/WeETH.sol]: thx

**Comment by seongyun-ko** [comment]
> @solipsis @seongyun-ko up to you guys if you think the extra 2000 gas on the `SLOAD` is worth it to apply the transfer blacklist to the sender as wel

yup we should

File: src/EETH.sol (hunk 1 above)
Comment (jtfirek) [reply src/EETH.sol:28]: Verified ✅

**Review by seongyun-ko — APPROVED** [review]
awesome LGTM

**Pull request merged**